src/core/storage/sframe_data/join_impl.cpp

```diff
@@ -156,67 +161,154 @@ hash_join_executor::hash_join_executor(const sframe &left,
     _right_join = true;
   }
 
+  ASSERT_EQ(_left_join_positions.size(), _right_join_positions.size());
+
+  /* handy script for reverse lookup */
+  for(size_t i = 0; i < _left_join_positions.size(); ++i) {
+    auto ret = _right_to_left_join_positions.emplace(_right_join_positions[i],
```

Comment:
I think `auto` can avoid some surprising implicit type conversions. Also, it makes code clear, brief and loosely coupled with a specific type, which can make refactoring code more easily.

For readability, the nowadays IDE can take advantage of type inference for `auto` to have a quick glance at the object's type.

Reply:
auto is good for readability for interesting derived types like iterators and stuff where the actual type of the object is "far away", or types may change due to overloading.
```
void do_something(std::map<std::string, size_t>& idx) {
   ...
    auto iter = idx.find(...)
}
```
where the alternative is kind of inconvenient
```
void do_something(std::map<std::string, size_t>& idx) {
    ...
    std::map<std::string, size_t>::iterator iter = idx.find(...) 
}
```
And also requires a lot of code changes if the argument types change in anyway.
```
void do_something(const std::map<std::string, size_t>& idx) {
    ...
    std::map<std::string, size_t>::iterator iter = idx.find(...)  // compile error. should be const_iterator
}
```

Also useful for things like range based for loops over complicated types (really just as above the "correct" type for a range based for loop is std::map<std::string, size_t>::value_type, that gets a little annoying.

But I will prefer that auto is not used for simple types, or potentially platform dependent types. 
```
uint64_t i = 0; // good. I can tell immediately what width this integer is and if it is signed.
auto i = 0; // bad. Hard to tell immediately.
float j = 0.0; //good. I know that this is a 32-bit float
auto j = 0.0; //bad. How wide is j?
```
This is crucial. For instance when we were porting from Linux/Mac to Windows, it was far easier to search for the "long" keyword to replace all uses of "long" with int64_t or "long long" (long is 32-bit on Windows, 64-bit elsewhere). Whereas if we had used "auto" this would have been significantly harder to fix.

Reply:
There are some uses of auto here for instance in integer for loops, which would be better to replace with a standard integer type like size_t, ssize_t, int64_t, or uint64_t.

Reply:
that's a good point. I will refrain from using auto for POD types.